src/module.c

```diff
@@ -233,30 +233,33 @@ void _MGraph_Query(void *args) {
 }
 
 void _MGraph_BulkInsert(void *args) {
+    // Establish thread-safe environment for batch insertion
     BulkInsertContext *context = (BulkInsertContext *)args;
-    Graph *g;           // Graph to populate.
-    size_t nodes = 0;   // Number of nodes created.
-    size_t edges = 0;   // Number of edge created.
+    RedisModuleCtx *ctx = RedisModule_GetThreadSafeContext(context->bc);
+    _MGraph_AcquireWriteLock(ctx);
+
     int argc = context->argc;
     RedisModuleString **argv = context->argv;
     RedisModuleString *rs_graph_name = argv[1];
-    RedisModuleCtx *ctx = RedisModule_GetThreadSafeContext(context->bc);
-    const char *graph_name = RedisModule_StringPtrLen(argv[1], NULL);
+    const char *graph_name = RedisModule_StringPtrLen(rs_graph_name, NULL);
+    size_t nodes = 0;   // Number of nodes created.
+    size_t edges = 0;   // Number of edge created.
 
-    _MGraph_AcquireWriteLock(ctx);
-    
     // Try to get graph, if graph does not exists create it.
-    g = Graph_Get(ctx, rs_graph_name);
-    if(g == NULL)
-        g = _MGraph_CreateGraph(ctx, rs_graph_name);
+    Graph *g = Graph_Get(ctx, rs_graph_name);
+    if (g == NULL) g = _MGraph_CreateGraph(ctx, rs_graph_name);
+
+    // Exit if graph creation failed
+    if (g == NULL) goto cleanup;
+
+    int rc = Bulk_Insert(ctx, argv+2, argc-2, g, graph_name, &nodes, &edges);
 
-    Bulk_Insert(ctx, argv+2, argc-2, g, graph_name, &nodes, &edges);
+    // Exit if insertion failed
+    if (rc == BULK_FAIL) goto cleanup;
```

Comment:
Yeah, this is a pretty thorny issue. In the ideal, I think that if a bulk insert cannot succeed completely, it should make no modifications at all, but that would be hard to enforce in the current setup.

As we update matrices and stores while processing tokens right now, we would need to roll back all updates in the event of failure, which is not even possible. Deleting properties from store schemas, for one, is not something we have an interface for, and as `argv` progresses we lose access to the strings that would have to be removed.

I think it would be better to verify that a bulk operation is viable before making any modifications at all, but that's not something we could do without moving CSV parsing into the module itself. The CLI connection has a token limit of ~1 million, so sufficiently large files will always have to be processed across multiple queries (making it inadequate for a single query to be internally consistent).